src/main/java/org/opensearch/flowframework/util/ParseUtils.java

```diff
@@ -225,10 +242,225 @@ public static Instant parseInstant(XContentParser parser) throws IOException {
+                function.run();
+            } else {
+                getWorkflow(requestedUser, workflowId, filterByEnabled, listener, function, client, clusterService, xContentRegistry);
+            }
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    /**
+     * Check if requested user has backend role required to access the resource
+     * @param requestedUser the user to execute the request
+     * @param resourceUser user of the resource
+     * @param workflowId workflow id
+     * @return boolean if the requested user has backend role required to access the resource
+     * @throws Exception exception
+     */
+    private static boolean checkUserPermissions(User requestedUser, User resourceUser, String workflowId) throws Exception {
+        if (resourceUser.getBackendRoles() == null || requestedUser.getBackendRoles() == null) {
+            return false;
+        }
+        // Check if requested user has backend role required to access the resource
+        for (String backendRole : requestedUser.getBackendRoles()) {
+            if (resourceUser.getBackendRoles().contains(backendRole)) {
+                logger.debug(
+                    "User: "
+                        + requestedUser.getName()
+                        + " has backend role: "
+                        + backendRole
+                        + " permissions to access config: "
+                        + workflowId
+                );
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if filter by backend roles is enabled and validate the requested user
+     * @param requestedUser the user to execute the request
+     */
+    public static void checkFilterByBackendRoles(User requestedUser) {
+        if (requestedUser == null) {
+            String errorMessage = "Filter by backend roles is enabled and User is null";
+            logger.error(errorMessage);
+            throw new FlowFrameworkException(errorMessage, RestStatus.BAD_REQUEST);
+        }
+        if (requestedUser.getBackendRoles().isEmpty()) {
+            String userErrorMessage = "Filter by backend roles is enabled, but User "
+                + requestedUser.getName()
+                + " does not have any backend roles configured";
+
+            logger.error(userErrorMessage);
+            throw new FlowFrameworkException(userErrorMessage, RestStatus.FORBIDDEN);
+        }
+    }
+
+    /**
+     * Get workflow
+     * @param requestUser the user to execute the request
+     * @param workflowId workflow id
+     * @param filterByEnabled filter by enabled setting
+     * @param listener action listener
+     * @param function workflow function
+     * @param client node client
+     * @param clusterService cluster service
+     * @param xContentRegistry contentRegister to parse get response
+     */
+    public static void getWorkflow(
```

Comment:
Non-blocking, but seems to me these methods should live outside of the ParseUtils class, since they dont have anything to do with parsing.

Reply:
Agreed. We don't actually parse here except the template but I wanted to keep it in a common file. Probably, will need another one of such.